pkg/scripts/install.go

```diff
@@ -13,177 +15,234 @@ import (
+func (s *UrlScriptHandler) Get(script string) *api.InstallResult {
+	if len(s.URL) == 0 {
+		return nil
 	}
-	if len(failedScripts) > 0 {
-		err = errors.NewInstallRequiredError(failedScripts, docker.ScriptsURLLabel)
+	scriptURL, err := url.ParseRequestURI(s.URL + "/" + script)
+	if err != nil {
+		glog.Infof("invalid script url %q: %v", s.URL, err)
+		return nil
+	}
+	return &api.InstallResult{
+		Script: script,
+		URL:    scriptURL.String(),
 	}
-
-	return
 }
 
-// InstallOptional downloads and installs a set of scripts into dstDir, the result is a
-// map of scripts with detailed information about each of the scripts install process
-func (i *installer) InstallOptional(scripts []string, dstDir string) []api.InstallResult {
-	return i.run(scripts, dstDir)
+// Install downloads the script and fix its permissions.
+func (s *UrlScriptHandler) Install(r *api.InstallResult) error {
+	downloadURL, err := url.Parse(r.URL)
+	if err != nil {
+		return err
+	}
+	dst := filepath.Join(s.DestinationDir, r.Script)
+	if _, err := s.download.Download(downloadURL, dst); err != nil {
+		if e, ok := err.(errors.Error); ok {
+			if e.ErrorCode == errors.ScriptsInsideImageError {
+				r.Installed = true
+				return nil
+			}
+		}
+		return err
+	}
+	if err := s.fs.Chmod(dst, 0755); err != nil {
+		return err
+	}
+	r.Installed = true
+	r.Downloaded = true
+	return nil
 }
 
-type downloadResult struct {
-	location string
-	err      error
+// SourceScriptHandler handles the case when the scripts are contained in the
+// source code directory.
+type SourceScriptHandler struct {
+	DestinationDir string
+	fs             util.FileSystem
 }
 
-func (i *installer) run(scripts []string, dstDir string) []api.InstallResult {
-	var userResults, sourceResults, defaultResults map[string]*downloadResult
-
-	// get scripts from user provided URL
-	if i.scriptsURL != "" {
-		userResults = i.download(i.scriptsURL, scripts, filepath.Join(dstDir, api.UserScripts))
+// Get verifies if the script is present in the source directory and get the
+// installation result.
+func (s *SourceScriptHandler) Get(script string) *api.InstallResult {
+	location := filepath.Join(s.DestinationDir, api.SourceScripts, script)
+	if s.fs.Exists(location) {
+		return &api.InstallResult{Script: script, URL: location}
 	}
-
-	// get scripts from source
-	sourceResults = make(map[string]*downloadResult, len(scripts))
-	for _, script := range scripts {
-		sourceResults[script] = &downloadResult{location: api.SourceScripts}
-		file := filepath.Join(dstDir, api.SourceScripts, script)
-		if !i.fs.Exists(file) {
-			sourceResults[script].err = errors.NewDownloadError(file, -1)
-		}
-		checkLocation := filepath.Join(dstDir, strings.TrimSuffix(api.SourceScripts, "/bin"), script)
-		if i.fs.Exists(checkLocation) {
-			glog.Warningf("The %q script found in .sti/%s, should be in .sti/bin/%s.", script, script, script)
-		}
+	location = strings.Replace(location, "s2i", "sti", 1)
```

Comment:
@soltysh will add comment and TODO here, but I don't want to introduce `api. DeprecatedSourceScripts` because 1) the name is super-ugly, 2) it is too much engineering for 1 character change.